### PR TITLE
8.0 New sourcing method "Use Existing bid"

### PR DESCRIPTION
**Pull request by yvaucher**

### CODE REVIEW

**Comment by gurneyalex** [comment]
red build on travis with flake8 issues, and at least one test crashing with

```
AttributeError: 'purchase.requisition' object has no attribute 'close_callforbids'
```

(/home/travis/build/OCA/vertical-ngo/framework_agreement_requisition/test/agreement_requisition.yml)

**Comment by yvaucher** [comment]
Fixed PEP8 errors, and rebased, the error on framework_agreement_requisition may be unrelated

**Comment by jgrandguillaume** [comment]
:+1:  Thanks ! Tested and it works fine !

File: logistic_requisition/model/logistic_requisition.py [hunk 1]
```diff
@@ -820,6 +827,36 @@ def _default_source_address(self):
...
+        if (self.sourcing_method == 'reuse_bid' and self.po_requisition_id and
+                self.proposed_product_id):
+            domain = [
+                ('order_id.requisition_id', '=', self.po_requisition_id.id),
+                ('product_id', '=', self.proposed_product_id.id),
+                ('order_id.state', '=', 'bid_selected'),
+                ('order_id.dest_address', '=', self.consignee_shipping_id.id),
+            ]
+            bid_lines = self.env['purchase.order.line'].search(domain)
+        self.selectable_bid_line_ids = (bid_lines.ids if bid_lines
```
Comment (lepistone) [inline logistic_requisition/model/logistic_requisition.py:860]: I'm unsure, maybe just assigning to `bid_lines` would just work? (I know it doesn't work in `search` and `write`, but maybe it does here.

File: logistic_requisition/model/logistic_requisition.py [hunk 2]
```diff
@@ -1245,3 +1312,27 @@ def onchange_selected_bid_id(self):
             purchase = self.selected_bid_id
             supplier_partner_id = purchase.partner_id.id
         self.supplier_partner_id = supplier_partner_id
+
+    @api.onchange('sourcing_method')
+    def onchange_sourcing_method(self):
+        """ Set `price_is` to fixed when reusing a bid
+        """
+        self.price_is = ('fixed' if self.sourcing_method == 'reuse_bid'
+                         else False)
```
Comment (lepistone) [inline logistic_requisition/model/logistic_requisition.py:1324]: are you sure about `else False`? That will empty the field in all other cases. Maybe you want to start with the `if`, and do nothing if `sourcing_method != 'reuse_bid'`.

File: logistic_requisition/tests/test_check_sourcing.py [hunk 3]
```diff
@@ -39,6 +39,12 @@ def test_procurement_sourcing_with_draft_pr_is_not_sourced(self):
         self.assertEquals(1, len(errors))
         self.assertIn('Purchase Requisition state should', errors[0])
 
+    def test_reuse_bid_sourcing_with_purchase_line_is_not_sourced(self):
```
Comment (lepistone) [inline logistic_requisition/tests/test_check_sourcing.py]: without?

**Comment by lepistone** [comment]
Minor remarks, but it looks great, thanks! :+1:

File: logistic_requisition/model/logistic_requisition.py (hunk 2 above)
Comment (yvaucher) [inline logistic_requisition/model/logistic_requisition.py:1324]: @lepistone This field needs to be set by user with other sourcing_methods
I want to empty it when we leave "reuse_bid" method. As I can't know what was previous value, I prefer to empty it in other case. It is a 2 options selection. That shouldn't be an overhead.

File: logistic_requisition/model/logistic_requisition.py [hunk 4]
```diff
@@ -981,6 +1029,17 @@ def _check_sourcing_procurement(self):
...
 
     @api.multi
+    def _check_sourcing_reuse_bid(self):
+        """Check sourcing for "reuse_bid" method.
+
+        :returns: list of error strings
+
+        """
+        if not self.selected_bid_line_id:
+            return ['{0}: No bid line selected'.format(self.name)]
```
Comment (guewen) [inline logistic_requisition/model/logistic_requisition.py]: Is it normal that it is not translated? (same question for the other `_check_sourcing*' methods)

File: logistic_requisition/model/purchase.py [hunk 5]
```diff
@@ -140,3 +140,14 @@ def unlink(self):
...
                     )
         return super(PurchaseOrderLine, self).unlink()
+
+    def copy_data(self, cr, uid, res_id, default=None, context=None):
+        """ When copying purchase order lines to reuse a bid
+        Link them to logistic requisition lines
+        """
+        res = super(PurchaseOrderLine, self).copy_data(
+            cr, uid, res_id, default=default, context=context)
+        if 'reuse_from_source' in context:
```
Comment (guewen) [inline logistic_requisition/model/purchase.py:153]: You should prevent a call on `None` context.

Comment (yvaucher) [inline logistic_requisition/model/purchase.py:153]: Sure, forgot it when I rewrote it in old api

File: logistic_requisition/model/logistic_requisition.py (hunk 4 above)
Comment (yvaucher) [inline logistic_requisition/model/logistic_requisition.py]: Did the same as L976-980, I guess both needs translations.
I will fix the others as well.

**Comment by gurneyalex** [comment]
:+1:

**Pull request merged**